File: npm/vite-plugin-ox-content/src/code-blocks.ts
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { promisify } from "node:util";
import { execFile } from "node:child_process";
import { importNapiModule } from "./napi";

const execFileAsync = promisify(execFile);

export interface ExtractedCodeBlock {
  language: string;
  meta: string;
  code: string;
  startLine: number;
  endLine: number;
}

export interface CodeBlockDiagnostic {
  ruleId: string;
  severity: "error" | "warning" | "info";
  message: string;
  line: number;
  column: number;
  endLine: number;
  endColumn: number;
  language?: string;
}

export interface CodeBlockLintOptions {
  /**
   * Languages to lint. Omit to lint every fenced block language.
   * @default undefined
   */
  languages?: string[];

  /**
   * Report fences without a language identifier.
   * @default false
   */
  requireLanguage?: boolean;

  /**
   * Report trailing whitespace in code block lines.
   * @default true
   */
  trailingSpaces?: boolean;
}

export interface DocsTestOptions {
  /**
   * Fence languages to collect as runnable examples.
   * @default ['js', 'jsx', 'ts', 'tsx', 'mjs', 'mts']
   */
  languages?: string[];

  /**
   * Require fence meta such as `test`, `runnable`, `vitest`, or `docs-test`.
   * @default true
   */
  requireMeta?: boolean;
}

export interface TypecheckCodeBlockOptions {
  /**
   * Fence languages to type-check.
   * @default ['ts', 'tsx']
   */
  languages?: string[];

  /**
   * Require fence meta such as `typecheck`, `twoslash`, or `typecheck=...`.
   * @default true
   */
  requireMeta?: boolean;

  /**
   * Command used to run the TypeScript checker.
   * @default 'tsgo'
   */
  tsgoCommand?: string;
}

export async function extractCodeBlocks(source: string): Promise<ExtractedCodeBlock[]> {
  const mod = await importNapiModule();
  return mod.extractCodeBlocks(source).map(normalizeBlock);
}

export async function lintCodeBlocks(
  source: string,
  options: CodeBlockLintOptions = {},
): Promise<CodeBlockDiagnostic[]> {
  const mod = await importNapiModule();
  return mod
    .lintCodeBlocks(source, {
      enabled: true,
      languages: options.languages,
      requireLanguage: options.requireLanguage,
      trailingSpaces: options.trailingSpaces,
    })
    .map(normalizeDiagnostic);
}

export async function extractDocsTests(
  source: string,
  options: DocsTestOptions = {},
): Promise<ExtractedCodeBlock[]> {
  const mod = await importNapiModule();
  return mod
    .extractDocsTests(source, {
      enabled: true,
      languages: options.languages,
      requireMeta: options.requireMeta,
    })
    .map(normalizeBlock);
}

export async function typecheckCodeBlocks(
  source: string,
  options: TypecheckCodeBlockOptions = {},
): Promise<CodeBlockDiagnostic[]> {
  if (!source.includes("```")) {
    return [];
  }

  const languages = new Set(
    (options.languages ?? ["ts", "tsx"]).map((language) => language.toLowerCase()),
  );
  const blocks = (await extractCodeBlocks(source)).filter((block) => {
    if (!languages.has(block.language.toLowerCase())) {
      return false;
    }
    return options.requireMeta === false || hasTypecheckMeta(block.meta);
  });
  if (blocks.length === 0) {
    return [];
  }

  const temp = await mkdtemp(join(tmpdir(), "ox-content-code-blocks-"));
  try {
    const files: string[] = [];
    await Promise.all(
      blocks.map(async (block, index) => {
        const extension = block.language.toLowerCase() === "tsx" ? "tsx" : "ts";
        const file = join(temp, `snippet-${index}.${extension}`);
        files.push(file);
        await writeFile(file, block.code);
      }),
    );

    try {
      await execFileAsync(
        options.tsgoCommand ?? "tsgo",
        ["--noEmit", "--pretty", "false", ...files],
        {
          cwd: process.cwd(),
          maxBuffer: 1024 * 1024 * 4,
        },
      );
      return [];
    } catch (error) {
      const output = commandOutput(error);
      return [
        {
          ruleId: "code-block-typecheck",
          severity: "error",
          message: output || "TypeScript code block type-checking failed.",
          line: blocks[0]?.startLine ?? 1,
          column: 1,
          endLine: blocks[0]?.startLine ?? 1,
          endColumn: 1,
          language: "ts",
        },
      ];
    }
  } finally {
    await rm(temp, { recursive: true, force: true });
  }
}

function hasTypecheckMeta(meta: string): boolean {
  return meta
    .split(/\s+/)
    .some(
      (token) => token === "typecheck" || token === "twoslash" || token.startsWith("typecheck="),
    );
}

function commandOutput(error: unknown): string {
  if (!error || typeof error !== "object") {
    return "";
  }
  const value = error as { stdout?: unknown; stderr?: unknown; message?: unknown };
  return [value.stdout, value.stderr, value.message]
    .filter((part): part is string => typeof part === "string" && part.trim().length > 0)
    .join("\n")
    .trim();
}

function normalizeBlock(block: {
  language: string;
  meta: string;
  code: string;
  startLine: number;
  endLine: number;
}): ExtractedCodeBlock {
  return block;
}

function normalizeDiagnostic(diagnostic: {
  ruleId: string;
  severity: string;
  message: string;
  line: number;
  column: number;
  endLine: number;
  endColumn: number;
  language?: string;
}): CodeBlockDiagnostic {
  return {
    ...diagnostic,
    severity:
      diagnostic.severity === "error" || diagnostic.severity === "info"
        ? diagnostic.severity
        : "warning",
  };
}

File: crates/ox_content_transform/src/features/code_blocks.rs
#![allow(clippy::redundant_pub_crate)]

use compact_str::CompactString;
use rustc_hash::FxHashSet;

use crate::{CodeBlockLintOptions, DocsTestOptions};

use super::segments::{is_closing_fence, parse_opening_fence};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ExtractedCodeBlock {
    pub language: String,
    pub meta: String,
    pub code: String,
    pub start_line: u32,
    pub end_line: u32,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CodeBlockDiagnostic {
    pub rule_id: String,
    pub severity: String,
    pub message: String,
    pub line: u32,
    pub column: u32,
    pub end_line: u32,
    pub end_column: u32,
    pub language: Option<String>,
}

pub fn extract_code_blocks(source: &str) -> Vec<ExtractedCodeBlock> {
    let mut blocks = Vec::new();
    let mut in_fence = false;
    let mut fence_char = b'\0';
    let mut fence_len = 0usize;
    let mut language = CompactString::default();
    let mut meta = CompactString::default();
    let mut code = String::new();
    let mut start_line = 0u32;
    let mut line_number = 0u32;

    for line in source.lines() {
        line_number += 1;
        if in_fence {
            if is_closing_fence(line, fence_char, fence_len) {
                blocks.push(ExtractedCodeBlock {
                    language: std::mem::take(&mut language).into_string(),
                    meta: std::mem::take(&mut meta).into_string(),
                    code: String::from(code.trim_end_matches('\n')),
                    start_line,
                    end_line: line_number.saturating_sub(1),
                });
                in_fence = false;
                fence_char = b'\0';
                fence_len = 0;
                code.clear();
            } else {
                code.push_str(line);
                code.push('\n');
            }
            continue;
        }

        if let Some(open) = parse_opening_fence(line) {
            in_fence = true;
            fence_char = open.fence_char;
            fence_len = open.fence_len;
            language = open.language;
            meta = open.meta;
            start_line = line_number + 1;
        }
    }

    blocks
}

pub fn lint_code_blocks(
    source: &str,
    options: Option<&CodeBlockLintOptions>,
) -> Vec<CodeBlockDiagnostic> {
    let Some(options) = options else {
        return Vec::new();
    };
    if options.enabled == Some(false) {
        return Vec::new();
    }

    let languages = options.languages.as_ref().map(|values| {
        values.iter().map(|value| value.to_ascii_lowercase()).collect::<FxHashSet<_>>()
    });
    let require_language = options.require_language.unwrap_or(false);
    let trailing_spaces = options.trailing_spaces.unwrap_or(true);
    let mut diagnostics = Vec::new();

    for block in extract_code_blocks(source) {
        if require_language && block.language.is_empty() {
            diagnostics.push(CodeBlockDiagnostic {
                rule_id: "code-block-language".to_string(),
                severity: "warning".to_string(),
                message: "Code block is missing a language identifier.".to_string(),
                line: block.start_line.saturating_sub(1),
                column: 1,
                end_line: block.start_line.saturating_sub(1),
                end_column: 1,
                language: None,
            });
        }

        if let Some(languages) = &languages {
            if !block.language.is_empty()
                && !languages.contains(&block.language.to_ascii_lowercase())
            {
                continue;
            }
        }

        if trailing_spaces {
            for (index, line) in block.code.lines().enumerate() {
                let trimmed = line.trim_end_matches([' ', '\t']);
                if trimmed.len() != line.len() {
                    diagnostics.push(CodeBlockDiagnostic {
                        rule_id: "code-block-trailing-spaces".to_string(),
                        severity: "warning".to_string(),
                        message: "Code block line has trailing whitespace.".to_string(),
                        line: block.start_line + index as u32,
                        column: (trimmed.len() + 1) as u32,
                        end_line: block.start_line + index as u32,
                        end_column: (line.len() + 1) as u32,
                        language: if block.language.is_empty() {
                            None
                        } else {
                            Some(block.language.clone())
                        },
                    });
                }
            }
        }
    }

    diagnostics
}

pub fn extract_docs_tests(
    source: &str,
    options: Option<&DocsTestOptions>,
) -> Vec<ExtractedCodeBlock> {
    let Some(options) = options else {
        return Vec::new();
    };
    if options.enabled == Some(false) {
        return Vec::new();
    }

    let languages = options.languages.as_ref().map_or_else(
        || ["js", "jsx", "ts", "tsx", "mjs", "mts"].into_iter().map(ToString::to_string).collect(),
        |values| values.iter().map(|value| value.to_ascii_lowercase()).collect::<FxHashSet<_>>(),
    );
    let require_meta = options.require_meta.unwrap_or(true);

    extract_code_blocks(source)
        .into_iter()
        .filter(|block| languages.contains(&block.language.to_ascii_lowercase()))
        .filter(|block| {
            !require_meta
                || block.meta.split_whitespace().any(|token| {
                    matches!(token, "test" | "runnable" | "vitest" | "docs-test")
                        || token.starts_with("test=")
                })
        })
        .collect()
}
